File: src/main/java/com/hileco/cortex/analysis/processors/DeadSwapProcessor.java
package com.hileco.cortex.analysis.processors;

import com.hileco.cortex.analysis.Graph;
import com.hileco.cortex.instructions.debug.NOOP;
import com.hileco.cortex.instructions.stack.SWAP;

public class DeadSwapProcessor implements Processor {

    @Override
    public void process(Graph graph) {
        graph.getGraphBlocks().forEach(graphBlock -> graphBlock.getGraphNodes().stream()
                .filter(graphNode -> graphNode.isInstruction(SWAP.class))
                .filter(swapNode -> {
                    var swap = (SWAP) swapNode.getInstruction().get();
                    return swap.getPositionLeft() == swap.getPositionRight();
                })
                .forEach(swapNode -> {
                    swapNode.getInstruction().set(new NOOP());
                }));
    }
}

File: src/main/java/com/hileco/cortex/analysis/processors/DeadSwapProcessor.kt
package com.hileco.cortex.analysis.processors

import com.hileco.cortex.analysis.Graph
import com.hileco.cortex.instructions.debug.NOOP
import com.hileco.cortex.instructions.stack.SWAP

class DeadSwapProcessor : Processor {
    override fun process(graph: Graph) {
        graph.graphBlocks.flatMap { it.graphNodes }
                .filter {
                    val instruction = it.instruction.get()
                    instruction is SWAP && instruction.topOffsetLeft == instruction.topOffsetRight
                }
                .forEach { it.instruction.set(NOOP()) }
    }
}

